### PR TITLE
Refactor execution-plan freeing mechanism

### PR DESCRIPTION
**Pull request by raz-mon**
This PR refactors the freeing mechanism of our execution-plan, and by thus allows for more complex execution-plans such as introduced in Foreach and Call {}.

### CODE REVIEW

File: src/execution_plan/execution_plan.c [hunk 1]
```diff
@@ -510,56 +517,88 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
+(
+	OpBase *opBase,     // operation to aggregate plans from
+	rax *plans          // plans map to insert to
+) {
+	ASSERT(plans != NULL);
 
-	return current_plan;
+	// add the plan if it doesn't exist already
+	ASSERT(sizeof(opBase->plan) == sizeof(uint64_t));
+	const uint64_t plan_ptr_be = _htobe64((uint64_t)opBase->plan);
```
Comment (raz-mon) [inline src/execution_plan/execution_plan.c:565]: Why (if at all) is this necessary?

File: src/execution_plan/execution_plan.c [hunk 2]
```diff
@@ -510,56 +517,88 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
-		if(prev_child_plan != child_plan && prev_child_plan != current_plan) {
-			_ExecutionPlan_FreeInternals(prev_child_plan);
-			prev_child_plan = child_plan;
-		}
+		_ExecutionPlan_FreeOpTree(op->children[i]);
 	}
 
 	// Free this op.
 	OpBase_Free(op);
+}
```
Comment (raz-mon) [inline src/execution_plan/execution_plan.c:555]: Function documentation

File: src/execution_plan/execution_plan.c [hunk 3]
```diff
@@ -510,56 +517,88 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
+		plan->record_pool = NULL;
 	}
 	if(plan->ast_segment != NULL) {
 		AST_Free(plan->ast_segment);
+		plan->ast_segment = NULL;
 	}
 	rm_free(plan);
 }
 
 // Free an op tree and its associated ExecutionPlan segments.
```
Comment (raz-mon) [inline src/execution_plan/execution_plan.c:542]: Doc fix: The segments are not freed here

**Review by raz-mon — COMMENTED** [review]
Great work
See few comments

File: src/execution_plan/execution_plan.c (hunk 1 above)
Comment (OfirMos) [reply src/execution_plan/execution_plan.c:566]: Do you mean why the _htobe64 is necessary?

File: src/execution_plan/execution_plan.c [hunk 4]
```diff
@@ -527,39 +534,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
+	if(opBase == NULL) {
+		return;
+	}
 
-	// Free each ExecutionPlan segment once all ops associated with it have been freed.
-	if(current_plan != child_plan) _ExecutionPlan_FreeInternals(child_plan);
+	// add the plan if it doesn't exist already
+	static_assert(sizeof(opBase->plan) == sizeof(uint64_t), "pointer size is not 64 bits");
+	const uint64_t plan_ptr_be = _htobe64((uint64_t)opBase->plan);
+	raxTryInsert(plans, (unsigned char *)&plan_ptr_be, sizeof(uint64_t), (void *)opBase->plan, NULL);
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c:566]: 1. consider switching from `rax` to `HashTable`
2. consider using the following approach:
```c
OpFree(op) {
  op->plan->ops_count--;
  if(op->plan->ops_count == 0) {
    _ExecutionPlan_FreeInternals(op->plan);
  }
}
```

File: src/execution_plan/execution_plan.c [hunk 5]
```diff
@@ -527,39 +534,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
 
-	return current_plan;
+	for(uint i = 0; i < opBase->childCount; i++) {
+		_ExecutionPlan_AggregatePlansFromOps(opBase->children[i], plans);
+	}
 }
 
+// free the execution plans and all of the operations
 void ExecutionPlan_Free(ExecutionPlan *plan) {
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c:574]:
```c
void ExecutionPlan_Free
(
    ExecutionPlan *plan
) {
```

File: src/execution_plan/execution_plan.c [hunk 6]
```diff
@@ -527,39 +534,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
-	if(plan == NULL) return;
+	if(plan == NULL) {
+		return;
+	}
+
+	// traverse the execution-plan graph (DAG -> no endless cycles), while
+	// aggregating the different execution-plans
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c:580]: while collecting the different/individual segments

File: src/execution_plan/execution_plan.c [hunk 7]
```diff
@@ -527,39 +534,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
+	rax *plans = raxNew();
+	const uint64_t plan_ptr_be = _htobe64((uint64_t)plan);
+	raxInsert(plans, (unsigned char *)&plan_ptr_be, sizeof(uint64_t), plan, NULL);
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c:583]: Redundant, please remove.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/execution_plan/execution_plan.c [hunk 8]
```diff
@@ -527,39 +539,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
+	// free op tree and collect execution-plans
+	// -------------------------------------------------------------------------
 
-	// Free each ExecutionPlan segment once all ops associated with it have been freed.
-	if(current_plan != child_plan) _ExecutionPlan_FreeInternals(child_plan);
+	// traverse the execution-plan graph (DAG -> no endless cycles), while
+	// while collecting the different segments, and freeing the op tree
+	dict *plans = HashTableCreate(&dt);
+	OpBase **ops = array_new(OpBase *, 1);
+	OpBase **ordered_ops = array_new(OpBase *, 1);
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c:561]:
```c
OpBase **to_visit = ...;
OpBase **visited = ..;
```

File: src/execution_plan/execution_plan.c [hunk 9]
```diff
@@ -527,39 +539,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
+	dictEntry *entry;
+	while(array_len(ops) > 0) {
+		op = array_pop(ops);
 
-	// Free all ops and ExecutionPlan segments.
-	_ExecutionPlan_FreeOpTree(plan->root);
+		// add the plan this op is affiliated with, if first met now
+		if((entry = HashTableAddRaw(plans, (void *)op->plan, NULL)) != NULL) {
+			HashTableSetVal(plans, entry, (void *)op->plan);
+		}
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c:573]: Please switch to: `HashTableAdd(plans, (void*)op->plan, (void*)op->plan);`

File: src/execution_plan/execution_plan.c [hunk 10]
```diff
@@ -527,39 +539,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
-	_ExecutionPlan_FreeInternals(plan);
-}
+		// add all direct children of op to ops
+		for(uint i = 0; i < op->childCount; i++) {
+			if(op->children[i] != NULL) {
+				array_append(ops, op->children[i]);
+			}
+		}
 
+		// free op
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c:582]: remove/update comment

File: src/execution_plan/execution_plan.c [hunk 11]
```diff
@@ -527,39 +539,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
+		array_append(ordered_ops, op);
+	}
+	while(array_len(ordered_ops) > 0) {
+		op = array_pop(ordered_ops);
+		OpBase_Free(op);
+	}
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c:588]:
```c
	for(int I = n; I >= 0; I--) {
 		op = visited[I];
 		OpBase_Free(op);
 	}
```

File: src/execution_plan/execution_plan.c [hunk 12]
```diff
@@ -527,39 +539,67 @@ static void _ExecutionPlan_FreeInternals(ExecutionPlan *plan) {
...
+	array_free(ordered_ops);
+	array_free(ops);
```
Comment (swilly22) [inline src/execution_plan/execution_plan.c:590]: swap rows

**Review by swilly22 — CHANGES_REQUESTED** [review]

**Review by swilly22 — APPROVED** [review]

**Pull request merged**